File: Day214_2025/Q214_GFG.java
import java.util.*;

class Q214_GFG {
    public int longestSubarray(int[] arr, int k) {
        int n = arr.length;
        int res = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);

        int sum = 0;
        for (int i = 0; i < n; i++) {
            if (arr[i] > k)
                sum++;
            else
                sum--;

            if (!map.containsKey(sum))
                map.put(sum, i);

            if (map.containsKey(sum - 1))
                res = Math.max(res, i - map.get(sum - 1));

            if (sum > 0)
                res = Math.max(res, i + 1);
        }
        return res;
    }
}

File: Day214_2025/Q214_GFG.py
class Solution:
    def longestSubarray(self, arr, k):
        n = len(arr)
        res = 0
        sum_ = 0
        index_map = {0: -1}

        for i in range(n):
            if arr[i] > k:
                sum_ += 1
            else:
                sum_ -= 1

            if sum_ not in index_map:
                index_map[sum_] = i

            if (sum_ - 1) in index_map:
                res = max(res, i - index_map[sum_ - 1])

            if sum_ > 0:
                res = max(res, i + 1)

        return res
